src/peergos/server/tests/simulation/Simulator.java

```diff
@@ -376,35 +410,72 @@ private void run(final Simulation simulation, final String user) {
                 return Optional.empty();
             }
         };
+        TriFunction<String, String, Boolean, Optional<Path>> randomSharedPathWithRetries = (owner, sharee, isDirectory) -> {
+            int keepTrying = 0;
+            while (keepTrying < 3) {
+                Optional<Path> path = randomSharedPath.apply(owner, sharee);
+                if (path.isPresent()) {
+                    Stat stat = fileSystems.getReferenceFileSystem(owner).stat(path.get());
+                    if (isDirectory && stat.fileProperties().isDirectory) {
+                        return path;
+                    } else if (! isDirectory && ! stat.fileProperties().isDirectory) {
+                        return path;
+                    }
+                }
+                keepTrying++;
+            }
+            return Optional.empty();
+        };
+        BiFunction<String, String, Optional<Path>> randomSharedDirectoryPath = (owner, sharee) ->
+             randomSharedPathWithRetries.apply(owner, sharee, true);
+
+        BiFunction<String, String, Optional<Path>> randomSharedFilePath = (owner, sharee) ->
+             randomSharedPathWithRetries.apply(owner, sharee, false);
 
 
         switch (simulation) {
             case READ_OWN_FILE:
                 Path readPath = randomFilePathForUser.get();
                 log(user, simulation, readPath);
-                read(user, readPath);
+                readFile(user, readPath);
                 break;
             case READ_SHARED_FILE:
-                Optional<Path> sharedOpt = randomSharedPath.apply(otherUser.get(), user);
+                Optional<Path> sharedOpt = randomSharedFilePath.apply(user, otherUser.get());
```

Comment:
Formerly this would result in `user` reading a (random) file that another (random)  user had shared with `user` -  is that still the case?

Reply:
Did you have a chance to check this?

Reply:
I'm still working through your comments. I get a fatal exception with the params in the original position. I hope to have time tomorrow to investigate further.

Reply:
Reversing the order of params like i did originally meant i could compare ref and peergos filesystems. 
As ref file system does not use sym links to model the shared dir structure i can't compare between the filesystems. That's fine i now just check to make sure the fie/dir exists in peergos dir.